pennylane/shadows/classical_shadow.py

```diff
@@ -0,0 +1,140 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Classical Shadows base class with processing functions"""
+import warnings
+from collections.abc import Iterable
+
+import pennylane.numpy as np
+import pennylane as qml
+from pennylane.shadows.utils import median_of_means, pauli_expval
+
+
+class ClassicalShadow:
+    """TODO: docstring"""
+
+    def __init__(self, bitstrings, recipes):
+        self.bitstrings = bitstrings
+        self.recipes = recipes
+
+        assert bitstrings.shape == recipes.shape
+        self.snapshots = len(bitstrings)
+
+        self.observables = [
+            qml.matrix(qml.PauliX(0)),
+            qml.matrix(qml.PauliY(0)),
+            qml.matrix(qml.PauliZ(0)),
+        ]
+
+    def local_snapshots(self, wires=None, snapshots=None):
+        r"""Compute the T x n x 2 x 2 local snapshots
+        i.e. compute :math:`3 U_i^\dagger |b_i \rangle \langle b_i| U_i - 1` for each qubit and each snapshot
+        """
+        bitstrings = self.bitstrings
+        recipes = self.recipes
+
+        if isinstance(snapshots, int):
+            pick_snapshots = np.random.choice(
+                np.arange(snapshots, dtype=np.int64), size=snapshots, replace=False
+            )
+            pick_snapshots = qml.math.convert_like(pick_snapshots, bitstrings)
+            bitstrings = qml.math.gather(bitstrings, pick_snapshots)
+            recipes = qml.math.gather(recipes, pick_snapshots)
+
+        if isinstance(wires, Iterable):
+            bitstrings = bitstrings[:, wires]
+            recipes = recipes[:, wires]
+
+        T, n = bitstrings.shape
+
+        U = np.empty((T, n, 2, 2), dtype="complex")
+        for i, u in enumerate(self.observables):
+            U[np.where(recipes == i)] = u
+
+        state = (
+            qml.math.cast((1 - 2 * bitstrings[:, :, None, None]), np.complex64) * U + np.eye(2)
+        ) / 2
+
+        return 3 * state - np.eye(2)[None, None, :, :]
+
+    @staticmethod
+    def _obtain_global_snapshots(local_snapshot):
+        global_snapshots = []
+        for T_snapshot in local_snapshot:
+            tensor_product = T_snapshot[0]
+            for n_snapshot in T_snapshot[1:]:
+                tensor_product = np.kron(tensor_product, n_snapshot)
+            global_snapshots.append(tensor_product)
+
+        return np.array(global_snapshots)
+
+    def global_snapshots(self, wires=None, snapshots=None):
+        """Compute the T x 2**n x 2**n global snapshots"""
+
+        local_snapshot = self.local_snapshots(wires, snapshots)
+
+        if local_snapshot.shape[1] > 16:
+            warnings.warn(
+                "Querying density matrices for n_wires > 16 is not recommended, operation will take a long time",
+                UserWarning,
+            )
+
+        return ClassicalShadow._obtain_global_snapshots(local_snapshot)
```

Comment:
Curiosity question: Is this abiding some standards or personal style preference to have the method call a staticmethod?

Reply:
I changed it mostly for the test, since I want to be able to monkeypatch the (expensive) state reconstruction if I don't need the result.

Reply:
I see! And is returning `ClassicalShadow._obtain_global_snapshots(local_snapshot)` instead of `self._obtain_global_snapshots(local_snapshot)` also style or does it have a purpose?

Reply:
Ah I didn't know you could use `self` to call a static method. I guess it's just style in this case? Although seeing this I agree that `self` is probably cleaner 😄.

Reply:
Changed it to call `self._obtain_global_snapshots`.